_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,507 @@
+  ...
+
+  @Autowired
+  private KafkaTemplate<String, User> userKafkaTemplate;
+
+  void sendCustomMessage(User user, String topicName) {
+    userKafkaTemplate.send(topicName, user);
+  }
+  ...
+}
+```
+
+As shown above, we can create `KafkaTemplate` for the specific type and send user records to designated topics.
+
+```java
+@Component
+class KafkaListenersExample {
+
+  @KafkaListener(topics = "reflectoring-user", groupId="reflectoring-user", containerFactory="userKafkaListenerContainerFactory")
+  void listener(User user) {
+    LOG.info("CustomUserListener [{}]", user);
+  }
+}
+```
+
+Since we have multiple listener containers, we are specifying which container factory (`containerFactory="userKafkaListenerContainerFactory"`) to be used while consuming records.
+
+If we don't specify this attribute it defaults to `kafkaListernerContainerFactory` which uses `StringSerializer` and `StringDeserializer` in our case.
+
+### Using Spring Messaging Message Conversion
+
+```java
+@Configuration
+class KafkaProducerConfig {
+  ...
+
+  @Bean
+  KafkaTemplate<String, String> kafkaTemplate() {
+    ...
+    kafkaTemplate.setMessageConverter(new StringJsonMessageConverter());
+    kafkaTemplate.setDefaultTopic("reflectoring-user");
+    return kafkaTemplate;
+  }
+  ...
+}
+```
+
+Spring Kafka provides `MessageConverter` abstractions with `JsonMessageConverter`. We can inject `MessageConverter` into our `KafkaTemplate` instance directly. We must configure a message converter that is compatible with the configured Kafka Serializer. We are using `StringJsonMessageConverter` with `StringSerializer`. Also, a default topic must to set, becaause we'll be separating records based on the type of data we are sending.
+
+```java
+@Configuration
+class KafkaConsumerConfig {
+  @Bean
+  public ConcurrentKafkaListenerContainerFactory<String, String> kafkaJsonListenerContainerFactory() {
+  ...
+  factory.setMessageConverter(new StringJsonMessageConverter());
+  return factory;
+  }
+}
+```
+
+```java
+@Component
+class KafkaListenersExample {
+  ...
+
+  @KafkaListener(topics = "reflectoring-user", groupId="reflectoring-user-mc", containerFactory="kafkaJsonListenerContainerFactory")
+  void listenerWithMessageConverter(User user) {
+    LOG.info("MessageConverterUserListener [{}]", user);
+  }
+}
+```
+
+Now in our `KafkaConsumerConfig`, we are creating a new `kafkaJsonListenerContainerFactory` with `StringJsonMessageConverter` set as our `MessageConverter`. We can inject this created listener factory directly to our `@KafkaListener` with `containerFactory="kafkaJsonListenerContainerFactory"` as shown in the above example.
+
+We can create any number of listeners for different types of records, without the need for multiple configurations for each type. When we use a `@KafkaListener`, the parameter type is provided to the method that will assist with the conversion (In our case the User type).
+
+```java
+@Component
+class KafkaSenderWithMessageConverter {
```

Comment:
Again, I don't know if this code example belongs to the text above or the text below